### PR TITLE
Implement schema checking for AVRO and JSON schemas

### PR DESCRIPTION
**Pull request by scrottty**
The current implementation doesn't allow for checking the schema of the message being sent to the topic. This stops messages being sent to topics configured to enforce the schema and enables message parsing for topics with a defined schema. 

This update allows the user to pass a `schemaType` and `schemaString` to the plugin. If the `schemaType` is not `AVRO` or `JSON`, it works as before. Otherwise a `GenericRecord` producer is created using the schema passed to the module. This checks the schema against the schema defined by the topic and checks message against that schema.

To get this to work I had to abstract a `Producer` object which holds the majority of the code that was previously in `Produce.java`. The `GenericProducer` and `ByteProducer` classes implement the code to create their respective producers and messages.

On the consumer side, the changes are much simpler. The messages are decoded using Avro tools to convert from bytes to a JSON object as per the schema before saving to the ION file

Some things to note:
* ~~I have tested this locally with an Avro schema and is working. But I haven't written any tests yet. Will wait for confirmation that this is a desired feature~~ Test have now been added
* Its been a long time writing strongly type code, and my first time in Java, so forgive me for any dumb coding here. Appreciate any feedback if there is a better way to do this
* ~~As with tests, formatting and documentation is yet to be done~~ Formatting and documentation have been added

An example call would look like:
```yaml
id: pulsar-test
namespace: test
tasks:
  - id: hello
    type: io.kestra.plugin.pulsar.Produce
    from: {
      "value": {
        "id": "test_id",
        "array": [1,2,3],
        "path":"/path/to/file"
      }
    }
    uri: pulsar://localhost:6650
    serializer: JSON
    topic: tet_topic
    schemaType: AVRO
    schemaString: "{\"type\": \"record\", \"name\": \"TestSchema\", \"fields\": [{\"name\": \"path\", \"type\": \"string\"}, {\"name\": \"id\", \"type\": \"string\"}, {\"name\": \"array\", \"type\": {\"type\": \"array\", \"items\": \"int\"}}]}"

```

### CODE REVIEW

File: src/main/java/io/kestra/plugin/pulsar/BaseProducer.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+import org.apache.pulsar.client.api.*;
+
+import io.kestra.core.models.executions.metrics.Counter;
+import io.kestra.core.runners.RunContext;
+import io.kestra.core.serializers.FileSerde;
+import io.reactivex.BackpressureStrategy;
+import io.reactivex.Flowable;
+
+public abstract class BaseProducer<T> {
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/pulsar/BaseProducer.java:27]:
```suggestion
public abstract class AbstractProducer<T> {
```

By convention, an abstract class is prefixed by Abstract.

File: src/main/java/io/kestra/plugin/pulsar/BaseProducer.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+  RunContext runContext;
+
+
+  public BaseProducer(RunContext runContext, PulsarClient client) {
+    this.client = client;
+    this.runContext = runContext;
+  }
+
+  public void ConstructProducer(
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/pulsar/BaseProducer.java:43]:
```suggestion
  public void constructProducer(
```

By convention, methods start with a lowercase (case should be lowerCamelCase)

Comment (loicmathieu) [reply src/main/java/io/kestra/plugin/pulsar/BaseProducer.java:43]: Please apply to all methods

File: src/main/java/io/kestra/plugin/pulsar/Produce.java [hunk 3]
```diff
@@ -146,156 +130,43 @@ public class Produce extends AbstractPulsarConnection implements RunnableTask<Pr
...
-            return ((LocalDateTime) timestamp).atZone(ZoneId.systemDefault()).toInstant().toEpochMilli();
-        }
+    @io.swagger.v3.oas.annotations.media.Schema(
+      title = "The topics schema type. If not AVRO or JSON, leave as byte",
+      description = "Required for connecting with topics using AVRO or JSON schemas"
+    )
+    @NotNull
+    @PluginProperty(dynamic = true)
+    @Builder.Default
+    private String schemaType = "byte";
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/pulsar/Produce.java:147]: Better to use an enum here with AVRO and JSON.
And null when there is no schema as byte here means no schema right?

Comment (scrottty) [reply src/main/java/io/kestra/plugin/pulsar/Produce.java:147]: yeh essentially. Will make that change

**Comment by scrottty** [comment]
> Hi, Sorry for the long delay to answer. Yes, this is a very welcome addition to our Pulsar plugin, thanks.
> 
> I requested a few changes and for the moment I didn't do a deep review but globally this seems good so please add test and documentation so we can proceed with this PR. It also need to be rebased as the Produce class changed since you implement this (if you struggle to rebase it, don't rebase I'll do it for you later).

@loicmathieu - Thank you for the review. I have made the changes as requested and added the documentation and tests. I also pushed changes for the consumer and trigger for when a topic has a defined schema

**Review by loicmathieu — APPROVED** [review]
LGTM, thanks for this nice addition.

I'll merge it as soon as CI will be green.

**Pull request merged**